ERCS/erc-7947.md

```diff
@@ -0,0 +1,171 @@
+     * MUST return `true` if the ownership recovery is successful.
+     * 
+     * @param newOwner the address of a new owner.
+     * @param provider the address of a recovery provider.
+     * @param proof an encoded proof of recovery (ZKP/ZKAI, signature, etc).
+     * @return `true` if recovery is successful, `false` (or revert) otherwise.
+     */
+    function recoverOwnership(
+        address newOwner,
+        address provider,
+        bytes memory proof
+    ) external returns (bool);
+}
+```
+
+A recovery provider MUST implement the following interface:
+
+```solidity
+/**
+ * @notice Defines a common recovery provider interface.
+ */
+interface IRecoveryProvider {
+    /**
+     * MUST be emitted in the `subscribe` function.
+     */
+    event AccountSubscribed(address indexed account);
+
+    /**
+     * MUST be emitted in the `unsubscribe` function.
+     */
+    event AccountUnsubscribed(address indexed account);
+
+    /**
+     * @notice A function that "subscribes" a smart account (msg.sender) to a recovery provider.
+     * SHOULD process and assign the `recoveryData` to the `msg.sender`.
+     * 
+     * @param recoveryData a recovery commitment (hash/ZKP public output) to be used 
+     * in the `recover` function to check a recovery proof validity.
+     */
+    function subscribe(bytes memory recoveryData) external;
+
+    /**
+     * @notice A function that revokes a smart account subscription.
+     * MUST delete all the recovery data associated with the `msg.sender`.
+     */
+    function unsubscribe() external;
+
+    /**
+     * @notice A function to get a recovery data (commitment) of an account.
+     * 
+     * @param account the account to get the recovery data of.
+     * @return the associated recovery data.
+     */
+    function getRecoveryData(address account) external view returns (bytes memory);
+
+    /**
+     * @notice A function that checks if a recovery of a smart account (msg.sender)
+     * to the `newOwner` is possible.
+     * SHOULD use `msg.sender`'s `recoveryData` to check the `proof` validity.
+     * MUST ensure that the `proof` can't be reused, e.g. update nonce.
+     * 
+     * @param newOwner the new owner to recover the `msg.sender` ownership to.
+     * @param proof the recovery proof.
+     */
+    function recover(address newOwner, bytes memory proof) external;
+}
+```
+
+## Rationale
+
+The AARI is expected to work with *any* account abstraction standard to allow for maximum account recovery flexibility. Whether it is [EIP-4337](./eip-4337.md) or [EIP-7702](./eip-7702.md), a particular smart account provider may support account recovery by simply implementing a common interface.
+
+Since the whole account recovery process is nothing but proving the knowledge of some alternative secret to a private key, it is essential for accounts to be able to "commit" to this secret. The `subscribe` function in the recovery provider interface allows precisely for that.
+
```

Comment:
Would you mind explaining why `unsubscribe` is important (in the document)? It isn't just for gas purposes, I assume?